spring-ws-core/src/main/java/org/springframework/ws/support/ObservationHelper.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2005-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.ws.support;
+
+import io.micrometer.common.util.internal.logging.WarnThenDebugLogger;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.ws.server.endpoint.observation.ObservationInterceptor;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.xml.sax.Attributes;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.DefaultHandler;
+
+import javax.xml.namespace.QName;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import javax.xml.transform.Source;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.sax.SAXSource;
+import javax.xml.transform.stream.StreamSource;
+import java.io.IOException;
+
+/**
+ * Helper class for observation tasks.
+ * @author Johan Kindgren
+ */
+public class ObservationHelper {
+
+    private final Log logger = LogFactory.getLog(getClass());
+
+    private static final WarnThenDebugLogger WARN_THEN_DEBUG_LOGGER = new WarnThenDebugLogger(ObservationInterceptor.class);
+    private static final QName UNKNOWN_Q_NAME = new QName("unknown", "unknow");
+
+    private final SAXParser saxParser;
+
+    public ObservationHelper() {
```

Comment:
Is there a reason to having multiple instances of this class? Could it be a singleton utility class instead?

Reply:
Or alternatively, could the methods be static and it not have a public constructor?

Reply:
It could certainly be a singleton, is something like this what you had in mind:
`java

   private static ObservationHelper INSTANCE;

    public static ObservationHelper getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new ObservationHelper();
        }
        return INSTANCE;
    }

    private ObservationHelper() {
`

Also noted that a minor bug slipped through the tests, I have a fix in place locally.

Reply:
Yes, that would be one option. I'll leave it up to @corneil to advise on what makes most sense given existing patterns used in this codebase.